xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-ui/src/main/resources/AnnotationCode/Script.xml

```diff
@@ -976,28 +976,239 @@ XWiki.Annotation = Class.create({
+      // a new range is created to recalculate the new ends.
+      var strictRange = this.fixRangeEndPoints(annDOMRange);
+
+      // All the nodes inside this range are extracted. Each node is wrapped inside the annotation marker span.
+      this.getTextNodesInRange(strictRange).forEach(textNode =&gt; this.markAnnotation(textNode, item));
+
+      // Add the marker span after the last span of this annotation.
+      var allSpans = this.annotatedElement.select('[class~=ID' + item.annotationId + ']');
+      if (allSpans.size() == 0) {
+        return;
+      }
+      var lastSpan = allSpans[allSpans.size() - 1];
+      // Create the annotation markers hidden by default, since annotations are added on the document hidden by default.
+      var markerSpan = new Element('span', {'id': 'ID' + item.annotationId, 'class' : 'hidden annotation-marker ' + item.state});
+      lastSpan.insert({after: markerSpan});
+      // Annotations are displayed on mouseover.
+      markerSpan.observe('click', this.onMarkerClick.bindAsEventListener(this, item.annotationId));
+    }.bind(this));
+  },
+
+  /**
+   * Surround this node with a span corresponding to it's annotation.
+   *
+   * @param markedNode the node that corresponds to the current annotation
+   * @param ann object holding information about the annotation
+  */
+  markAnnotation: function(markedNode, ann) {
+    var wrapper = document.createElement('span');
+    wrapper.addClassName('annotation');
+    wrapper.addClassName('ID' + ann.annotationId);
+
+    var parentNode = markedNode.parentElement;
+    parentNode.replaceChild(wrapper, markedNode);
+    wrapper.appendChild(markedNode);
+  },
+
+  /**
+   * For the first and last node, split the nodes at the known offset to not annotate the whole text. Create a new range
+   * with these new nodes.
+   *
+   * @param range the DOM range of the annotated text
+  */
+  fixRangeEndPoints: function(range) {
+    var strictRange = new Range();
+
+    // In case the range starts and ends in the same text node, the first split could alter the length of the node and
+    // break the second split.
+    var endOffset = range.endOffset;
+    if (range.startContainer == range.endContainer) {
+      endOffset -= range.startOffset;
+    }
+
+    var newStartNode = range.startContainer;
+    // The split is done only if the offset is not before first or after last character for not creating empty text nodes.
+    if (range.startOffset &gt; 0 &amp;&amp; range.startOffset &lt; (range.startContainer.length - 1)) {
+      newStartNode = range.startContainer.splitText(range.startOffset);
+    }
+    strictRange.setStartBefore(newStartNode);
+
+    var newEndNode = range.endContainer;
+    // The split is done only if the offset is not before first or after last character for not creating empty text nodes.
+    if (endOffset &gt; 0 &amp;&amp; endOffset &lt; (range.endContainer.length - 1)) {
+      newEndNode = range.endContainer.splitText(endOffset + 1);
```

Comment:
I think I misunderstood the meaning of the `Range.endOffset`. In my case, `range.endOffset` is the index of the last character that should be included in the annotation. If I have `foo` annotated, then `endOffset = 2` and `splitText(2)` will put compute the nodes `fo` and `o` 
Because of this `endOffset + 1` I added also the check of `endOffset < (range.endContainer.length - 1)`

But after checking again the documentation, in the above example the `endOffset` should have been 3. So the problem could be solved  by adding here a `+1` https://github.com/xwiki/xwiki-platform/pull/1700/files#diff-6fae9d69c9935f88c56cf7c9fde0f2f98b2cee09a3810b706400d3c1b007f437R1099  (I think Script.xml diff should be expanded for the link above to work, it's not done by default)
@mflorea WDYT?